File: figs/oplog_in_practice.lua
package.path = package.path .. ';lib/?.lua'
local luamp = require 'luamp'
local stream = require 'stream'
local table = require 'table'
local string = require 'string'

local SEG_WIDTH = 1
local SEG_HEIGHT = 0.6
local HEADER_WIDTH = 1.2
local FOOTER_WIDTH = 1.2

local matrix = luamp.layouts.matrix(luamp.origin, 0.8, 1,
    {{false, false},
     {false, false},
     {false, false}})

local figs = {}

stream.from_list({
    luamp.text(
        matrix[1][1],
        luamp.directions.center,
        '$\\text{epoch}_0$'),
    luamp.text(
        matrix[3][1],
        luamp.directions.center,
        '$\\text{epoch}_m$'),
    luamp.text(
        matrix[2][2],
        luamp.directions.right,
        '\\dots')})
    :flatten()
    :collect(figs)

local function epoch0()
    local texts = {'op', '\\dots', 'idx', 'op', '\\dots', 'idx'}
    local entire_width = HEADER_WIDTH + FOOTER_WIDTH + SEG_WIDTH * #texts
    local entire_height = SEG_HEIGHT
    local leftmost = matrix[1][2]
    local rightmost = leftmost + luamp.point(entire_width, 0)

    stream.from_list({
        stream.iterate(
            leftmost + luamp.point(HEADER_WIDTH, 0),
            function(x)
                return x + luamp.point(SEG_WIDTH, 0)
            end)
            :map(function(x)
                return luamp.line(
                    x + luamp.point(0, SEG_HEIGHT/2),
                    x + luamp.point(0, -SEG_HEIGHT/2))
            end)
            :take(#texts + 1),
        stream.zip(
            stream.from_list(texts),
            stream.iterate(
                leftmost + luamp.point(HEADER_WIDTH + SEG_WIDTH/2, 0),
                function(x)
                    return x + luamp.point(SEG_WIDTH, 0)
                end))
            :map(function(x)
                local text = x[1]
                local center = x[2]
                return luamp.text(center, luamp.directions.center, text)
            end),
        luamp.text(
            leftmost + luamp.point(HEADER_WIDTH/2, 0),
            luamp.directions.center,
            'header'),
        luamp.text(
            rightmost + luamp.point(-FOOTER_WIDTH/2, 0),
            luamp.directions.center,
            'footer'),
        luamp.rectangle(luamp.centroid(leftmost, rightmost), entire_width, entire_height)})
        :flatten()
        :collect(figs)
end
epoch0()

local function epoch_m()
    local entire_width = HEADER_WIDTH + 2 * SEG_WIDTH
    local entire_height = SEG_HEIGHT
    local leftmost = matrix[3][2]
    local rightmost = leftmost + luamp.point(entire_width, 0)

    stream.from_list({
        stream.iterate(
            leftmost + luamp.point(HEADER_WIDTH, 0),
            function(x)
                return x + luamp.point(SEG_WIDTH, 0)
            end)
            :map(function(x)
                return luamp.line(
                    x + luamp.point(0, SEG_HEIGHT/2),
                    x + luamp.point(0, -SEG_HEIGHT/2))
            end)
            :take(3),
        stream.zip(
            stream.from_list({'$\\text{op}_{m,0}$', '\\dots'}),
            stream.iterate(
                leftmost + luamp.point(HEADER_WIDTH + SEG_WIDTH/2, 0),
                function(x)
                    return x + luamp.point(SEG_WIDTH, 0)
                end))
            :map(function(x)
                local text = x[1]
                local center = x[2]
                return luamp.text(center, luamp.directions.center, text)
            end),
        luamp.text(
            leftmost + luamp.point(HEADER_WIDTH/2, 0),
            luamp.directions.center,
            'header'),
        luamp.arrow(
            rightmost + luamp.point(SEG_WIDTH * 3 / 2, 0) + luamp.point(-0.3, 0),
            rightmost + luamp.point(SEG_WIDTH/2, 0)),
        luamp.text(
            rightmost + luamp.point(SEG_WIDTH * 3 / 2, 0) + luamp.point(0.2, 0),
            luamp.directions.center,
            '$\\text{op}_{m,i}$'),
        luamp.rectangle(luamp.centroid(leftmost, rightmost), entire_width, entire_height)})
        :flatten()
        :collect(figs)

    local vs = {
        luamp.point(0, SEG_HEIGHT/2),
        luamp.point(SEG_WIDTH, SEG_HEIGHT/2),
        luamp.point(SEG_WIDTH, -SEG_HEIGHT/2),
        luamp.point(0, -SEG_HEIGHT/2)}
    local vs = stream.from_list(vs)
        :map(function(x)
            return rightmost + x
        end)
        :collect()
    stream
        .zip(
            stream.from_list(vs),
            stream.from_list(vs)
                :drop(1))
        :map(function(x)
            return luamp.line(x[1], x[2], {line_style=luamp.line_styles.dashed})
        end)
        :collect(figs)
end
epoch_m()

print(luamp.figure(table.unpack(figs)))

File: figs/oplog_in_practice.py
from fathom import Point, ORIGIN
from fathom.tikz import Canvas
import fathom.geometry as geo
import fathom.layout as layout
import fathom.tikz.colors as colors
import fathom.tikz.line_styles as line_styles
import fathom.tikz.locations as locations
from itertools import *

SEG_WIDTH = 1
SEG_HEIGHT = 0.6
HEADER_WIDTH = 1.2
FOOTER_WIDTH = 1.2


def epoch0(canvas, leftmost):
    texts = ['header', 'op', '\\dots', 'idx', 'op', '\\dots', 'idx', 'footer']
    rects = repeat(leftmost + Point(HEADER_WIDTH, 0))
    rects = accumulate(rects, lambda x, _: x + Point(SEG_WIDTH, 0))
    rects = islice(rects, len(texts) - 2)
    rects = [geo.Rectangle(center=x + Point(SEG_WIDTH/2, 0),
                           width=SEG_WIDTH, height=SEG_HEIGHT) for x in rects]
    rects.insert(0, geo.Rectangle(
        center=leftmost + Point(HEADER_WIDTH / 2, 0),
        width=HEADER_WIDTH,
        height=SEG_HEIGHT))
    rects.append(geo.Rectangle(
        center=rects[-1].center() + Point(SEG_WIDTH / 2 + FOOTER_WIDTH / 2, 0),
        width=FOOTER_WIDTH,
        height=SEG_HEIGHT))

    for x, t in zip(rects, texts):
        canvas.new_text(anchor=x.center(), text=t)

    canvas.new_rectangle(
        vertices=[
            rects[0].vertices()[0],
            rects[-1].vertices()[1],
            rects[-1].vertices()[2],
            rects[0].vertices()[3]])

    for x in rects[1:]:
        canvas.new_line(src=x.vertices()[0], dst=x.vertices()[3])


def epoch_m(canvas, leftmost):
    texts = ['header', 'op', '\\dots']
    rects = repeat(leftmost + Point(HEADER_WIDTH, 0))
    rects = accumulate(rects, lambda x, _: x + Point(SEG_WIDTH, 0))
    rects = islice(rects, len(texts) + 1)
    rects = [geo.Rectangle(center=x + Point(SEG_WIDTH/2, 0),
                           width=SEG_WIDTH, height=SEG_HEIGHT) for x in rects]
    rects.insert(0, geo.Rectangle(
        center=leftmost + Point(HEADER_WIDTH / 2, 0),
        width=HEADER_WIDTH,
        height=SEG_HEIGHT))

    for x, t in zip(rects, texts):
        canvas.new_text(anchor=x.center(), text=t)

    canvas.new_rectangle(
        vertices=[
            rects[0].vertices()[0],
            rects[2].vertices()[1],
            rects[2].vertices()[2],
            rects[0].vertices()[3]
        ]
    )
    for x in rects[1:3]:
        canvas.new_line(src=x.vertices()[0], dst=x.vertices()[3])

    v_srcs = rects[3].vertices()
    v_dsts = v_srcs[1:]
    for s, t in zip(v_srcs, v_dsts):
        canvas.new_line(src=s, dst=t, line_style=line_styles.DASHED)

    canvas.new_arrow(src=rects[-1].center(), dst=rects[-2].center())
    canvas.new_text(anchor=rects[-1].center(),
                    text='$\\text{op}_{m,i}$', location=locations.EAST)

if __name__ == '__main__':
    matrix = layout.matrix(
        h_sep=1,
        v_sep=0.8,
        n_rows=3,
        n_cols=2,
        top_left=ORIGIN)

    canvas = Canvas(
        preamble=['\\usepackage{amsmath}'],
        leading_instructions=['\\footnotesize'])

    canvas.new_text(
        anchor=matrix[0][0],
        text='$\\text{epoch}_0$')
    canvas.new_text(
        anchor=matrix[1][1],
        text='\\normalsize\\dots',
        location=locations.EAST)
    canvas.new_text(
        anchor=matrix[2][0],
        text='$\\text{epoch}_m$')

    epoch0(canvas, matrix[0][1])
    epoch_m(canvas, matrix[2][1])

    print(canvas.draw())
